File: leetcodeProblems/16.3-sum-closest.js
/*
 * @lc app=leetcode id=16 lang=javascript
 *
 * [16] 3Sum Closest
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
	let diff = Number.MAX_SAFE_INTEGER + 1;
	nums.sort((a, b) => a - b);
	let n = nums.length;
	for (let i = 0; i < n; i++) {
		let low = i + 1;
		let high = n - 1;
		while (low < high) {
			let sum = nums[i] + nums[low] + nums[high];
			if (Math.abs(target - sum) < Math.abs(diff))
				diff = target - sum;
			if (target > sum)
				low++;
			else
				high--;
		}
	}
	return target - diff;
};
// @lc code=end

console.log(threeSumClosest([-1, 2, 1, -4], 0));

File: leetcodeProblems/16.3-sum-closest.cpp
#include <bits/stdc++.h>

using namespace std;

int     threeSumCloset(vector<int> nums, int target)
{
    sort(nums.begin(), nums.end());
    int diff = INT_MAX;
    int n = nums.size();
    for (int i = 0; i < n; i++) {
        int low = i + 1;
        int high = n - 1;
        while (low < high) {
            int sum = nums[i] + nums[low] + nums[high];
            if (abs(target - sum) < abs(diff))
                diff = target - sum;
            if (target > sum)
                low++;
            else
                high--;
        }
    }
    return target - diff;
}

int     main(void)
{
    vector<int> arr;
    arr.push_back(-1);
    arr.push_back(2);
    arr.push_back(1);
    arr.push_back(-4);
    cout << threeSumCloset(arr, 1) << endl;
    return (0);
}
